ocaml/xapi/xapi_host.ml

```diff
@@ -518,6 +518,33 @@ let reboot ~__context ~host =
   shutdown_and_reboot_common ~__context ~host "Host is rebooting" "Host is rebooting"
 	`shutdown "/sbin/shutdown -r now"
 
+let read_lines fi =
+	let rec loop acc =
+		let l = try Some (input_line fi) with _ -> None
+		in match l with
+			| None -> List.rev acc
+			| Some l -> loop (l :: acc)
+	in loop []
+
+let read_file file =
+	let fi = try Some (open_in file) with _ -> None in
+	match fi with
+	| None -> []
+	| Some fi ->
+		let lines = read_lines fi in
+		close_in fi ;
+		lines
+
+let reboot_required ~__context ~host =
+  let xen_file_count = try Array.length (Sys.readdir "/run/reboot-required.d/xen/") with _ -> 0 in
+  let kernel_file_count = try Array.length (Sys.readdir "/run/reboot-required.d/kernel/") with _ -> 0 in
+  let required = xen_file_count > 0 || kernel_file_count > 0 in
```

Comment:
_Any_ package should be able to write a file under /run/reboot-required.d/<pkgname>/... rather than being a xen and kernel specific mechanism.